app/celery/letters_pdf_tasks.py

```diff
@@ -262,6 +265,100 @@ def process_virus_scan_passed(self, filename):
         update_notification_status_by_id(notification.id, NOTIFICATION_TECHNICAL_FAILURE)
 
 
+@notify_celery.task(bind=True, name='sanitise-letter', max_retries=15, default_retry_delay=300)
+def sanitise_letter(self, filename):
+    try:
+        reference = get_reference_from_filename(filename)
+        notification = dao_get_notification_by_reference(reference)
+
+        current_app.logger.info('Notification ID {} Virus scan passed: {}'.format(notification.id, filename))
+
+        if notification.status != NOTIFICATION_PENDING_VIRUS_CHECK:
+            current_app.logger.exception('Sanitise letter called for notification {} which has is in {} state'.format(
+                notification.id, notification.status))
+            return
+
+        notify_celery.send_task(
+            name=TaskNames.SANITISE_LETTER,
+            kwargs={
+                'notification_id': str(notification.id),
+                'filename': filename,
+            },
+            queue=QueueNames.TEMPLATE_PREVIEW,
+        )
+    except Exception:
+        try:
+            current_app.logger.exception(
+                "RETRY: calling sanitise_letter task for notification {} failed".format(notification.id)
+            )
+            self.retry(queue=QueueNames.RETRY)
+        except self.MaxRetriesExceededError:
+            message = "RETRY FAILED: Max retries reached. " \
+                      "The task sanitise_letter failed for notification {}. " \
+                      "Notification has been updated to technical-failure".format(notification.id)
+            update_notification_status_by_id(notification.id, NOTIFICATION_TECHNICAL_FAILURE)
```

Comment:
This is fine - except when the template preview app throws an exception but it's actually a file that we can not validate. Until we fix template-preview so that it doesn't throw exceptions we will set letters to tech-failure when it's actually invalid.

Reply:
We catch all exceptions from template preview and mark the notification as validation-failed, so this shouldn't be an issue. We were already doing that in an error handler, and in my template preview PR I've kept that behaviour but moved it to the function that is sanitising the letter - https://github.com/alphagov/notifications-template-preview/pull/399/files#diff-cc3699ca803f977365cdb0525d419092R123.